src/reason_pprint_ast.ml

```diff
@@ -2996,94 +3092,142 @@ class printer  ()= object(self:'self)
+      if higherPrecedenceThan shiftPrecedence withPrecedence then rightRecurse
+      else if (higherPrecedenceThan withPrecedence shiftPrecedence) then
+        formatPrecedence ~loc:reducesAfterRight.pexp_loc rightRecurse
+      else (
+        if isRightAssociative withPrecedence then
+          rightRecurse
+        else
+          formatPrecedence ~loc:reducesAfterRight.pexp_loc rightRecurse
+      )
+    | VeryHighPrecedence itms ->
+      formatAttachmentApplication applicationFinalWrapping None (itms, Some reducesAfterRight.pexp_loc)
+
+  method ensureExpression expr ~reducesOnToken =
+    match self#unparseExprRecurse expr with
+    | SpecificPrecedence ({reducePrecedence; shiftPrecedence}, leftRecurse) ->
+      if higherPrecedenceThan reducePrecedence reducesOnToken then leftRecurse
+      else if higherPrecedenceThan reducesOnToken reducePrecedence then
+        formatPrecedence ~loc:expr.pexp_loc leftRecurse
+      else (
+        if isLeftAssociative reducesOnToken then
+          leftRecurse
+        else
+          formatPrecedence ~loc:expr.pexp_loc leftRecurse
+      )
+    | VeryHighPrecedence itms -> formatAttachmentApplication applicationFinalWrapping None (itms, Some expr.pexp_loc)
 
-  (*
-    Returns the breakable list of terms for function or (todo) constructor
-    application:
-
-    Should essentially return the list of formatted items that can be treated
-     "simply" (not necessarily "simple" expressions).
-
-     Two items:
-
-       let x = myFun (Red 10);
-
-     Still two items:
-
-       let x = Red 10;
 
-     Three items
+  (** Attempts to unparse: The beginning of a more general printing algorithm,
+      that determines how to print based on precedence of tokens and rules.
+      The end goal is that this should be completely auto-generated from the
+      Menhir parsing tables. We could move more and more into this function.
 
-       let x = Red 10 20;
+      You could always just call self#expression, but `unparseExpr` will render
+      infix/prefix/unary/terary fixities in their beautiful forms while
+      minimizing parenthesis.
   *)
+  method unparseExpr x =
+    match self#unparseExprRecurse x with
+    | SpecificPrecedence ({reducePrecedence; shiftPrecedence}, leftRecurse) -> leftRecurse
+    | VeryHighPrecedence itms -> formatAttachmentApplication applicationFinalWrapping None (itms, Some x.pexp_loc)
 
-  method arraySugarApplicationItems x =
-    match x.pexp_desc with
-      | Pexp_apply (e, l) -> (
-          match self#sugar_expr x with
-            | Some s -> Some s
-            | None -> None
-      )
-      | _ -> None
+  method unparseExprApplicationItems x =
+    match self#unparseExprRecurse x with
+    | SpecificPrecedence ({reducePrecedence; shiftPrecedence}, leftRecurse) -> ([leftRecurse], Some x.pexp_loc)
+    | VeryHighPrecedence itms -> (itms, Some x.pexp_loc)
 
-  method consecutiveInfixApplicationItems x =
+  method unparseExprRecurse x =
```

Comment:
This is now the main entrypoint to printing expressions. It falls through to `self#expression` if the special "unparser" can't handle that kind of expression yet.